Add __array_ufunc__ to Series / Array

jorisvandenbossche opened the pull request:
This PR:

* adds a basic (but incomplete) `__array_ufunc__` implementation for `IntegerArray` (to be able to check it is correctly used from Series for ExtensionArrays)
* adds `Series.__array_ufunc__` (not yet for `Index`)

---

@TomAugspurger revived my branch, will try to work on it a bit further this afternoon

What this already does is a basic implementation of the protocol for IntegerArray for simple ufuncs (call) for all IntegerArrays, and Series dispatching to the underlying values.

One question is if we want to force EA authors to implement an `__array_ufunc__` (eg by having a default implementation returning NotImplemented).

jorisvandenbossche committed:
Add __array_ufunc__ to Series / Array

pep8speaks commented:
Hello @jorisvandenbossche! Thanks for submitting the PR.

 - There are no PEP8 issues in the file [`pandas/core/arrays/integer.py`](https://github.com/jorisvandenbossche/pandas/blob/2fffac3b464da3db1e1931c18ccba886d67d6564/pandas/core/arrays/integer.py) !

 - There are no PEP8 issues in the file [`pandas/core/series.py`](https://github.com/jorisvandenbossche/pandas/blob/2fffac3b464da3db1e1931c18ccba886d67d6564/pandas/core/series.py) !

 - There are no PEP8 issues in the file [`pandas/tests/arrays/test_integer.py`](https://github.com/jorisvandenbossche/pandas/blob/2fffac3b464da3db1e1931c18ccba886d67d6564/pandas/tests/arrays/test_integer.py) !

TomAugspurger commented on pandas/core/arrays/integer.py:
```diff
@@ -293,6 +294,33 @@ def __array__(self, dtype=None):
+                else:
+                    result[mask] = np.nan
+                    return result
+
```
We should be able to handle binary ufuncs like `np.add(IntegerArray, 2)` right? Do you want to do that here?

TomAugspurger replied:
Ah, I see you have a test for this.

jorisvandenbossche replied:
Yes, for sure we should handle that. I didn't have any test for that yet, only for `np.add(IntegerArray, IntegerArray)`

Basically what I did up to now was a basic implemention on IntegerArray, just to be able to test the machinery on Series as well.

jorisvandenbossche replied:
This is handled by the `maybe_dispatch_ufunc_to_dunder_op` now

TomAugspurger reviewed (commented):
> One question is if we want to force EA authors to implement an __array_ufunc__ (eg by having a default implementation returning NotImplemented).

Would the motivation here be to simplify pandas internal code? We can skip any checks for it?

jorisvandenbossche commented:
BTW, if you also already had some stuff on this, feel free to push the branch if I could use something of it

jorisvandenbossche commented:
Considering the IntegerArray ``__array_ufunc__`` implementation, the question comes up to what extent do we want to integrate our `_create_arithmetic_method` with it. 
Because the `__array_ufunc__` needs to be able to do everything that  `_create_arithmetic_method` does. So do we want to use the dunder methods created in `_create_arithmetic_method` inside `__array_ufunc__`, or do we want to base the ops dunder methods on the `__array_ufunc__` (like the `NDArrayOperatorsMixin` does: https://docs.scipy.org/doc/numpy-1.15.1/reference/generated/numpy.lib.mixins.NDArrayOperatorsMixin.html#numpy.lib.mixins.NDArrayOperatorsMixin)

jorisvandenbossche commented:
@TomAugspurger the first option (use `__add__` inside `__array_ufunc__(np.add, ..)` instead of the other way around) is basically what you did for SparseArray:

https://github.com/pandas-dev/pandas/blob/5a0c91e5a5bc61c71da08d6968d973255829678d/pandas/core/arrays/sparse.py#L1450-L1487

jorisvandenbossche committed:
expand IntegerArray.__array_ufunc__

jorisvandenbossche committed:
fix Series.__array_ufunc__ and consolidate dispatch

jorisvandenbossche committed:
test Series array_ufunc fallback to numpy array for DecimalArray

jorisvandenbossche committed:
fix import

TomAugspurger commented on pandas/core/series.py:
```diff
@@ -623,6 +623,29 @@ def view(self, dtype=None):
 
+    def __array_ufunc__(self, ufunc, method, *inputs, **kwargs):
+        inputs = tuple(
+            x._values if isinstance(x, type(self)) else x
```
nitpick: could move `type(self)` outside this expression, so that it's only evaluated once.

TomAugspurger commented on pandas/core/series.py:
```diff
@@ -623,6 +623,29 @@ def view(self, dtype=None):
         return self._constructor(self._values.view(dtype),
                                  index=self.index).__finalize__(self)
 
+    def __array_ufunc__(self, ufunc, method, *inputs, **kwargs):
```
(How) should we handle the `out` parameter? Is that possible? On my WIP I had

```diff
+        if out:
+            kwargs['out'] = tuple(x._values if isinstance(x, cls) else x
+                                  for x in out)
```

but I don't know if that was tested.

jorisvandenbossche replied:
xarray is for example simply erroring in such a case:

https://github.com/pydata/xarray/blob/5ebed791db006075701c16b2e8eb9566207f2548/xarray/core/arithmetic.py#L53-L60

In the above, you would pass through an ExtensionArray as `out`, but that means that the ExtensionArray implementation of `out` should support it. 
But, maybe we should indeed do pass it through here (as you show), and let the `ExtensionArray.__array_ufunc__` error on an EA as `out` if it wants?

TomAugspurger replied:
Would it necessarily be an ExtensionArray, or could it also be a Series for `out`?

But raising seems fine for now.

jorisvandenbossche replied:
In master, an `out` argument that is not an array does not seem to be supported:

```
In [47]: s = pd.Series([1, 2, 3])

In [49]: out = s.copy()

In [50]: np.add(s, s, out=out)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-50-482c18e97629> in <module>()
----> 1 np.add(s, s, out=out)

TypeError: return arrays must be of ArrayType
```

TomAugspurger commented:
cc @shoyer if you have thoughts here as well.

shoyer commented on pandas/core/arrays/integer.py:
```diff
@@ -293,6 +295,44 @@ def __array__(self, dtype=None):
+            np.array(x) if isinstance(x, type(self)) else x
+            for x in inputs
+        )
+        return np.array(self).__array_ufunc__(
```
You probably want to call the public ufunc again here, e.g., `return getattr(ufunc, method)(*inputs, **kwargs)`. Otherwise you don't invoke the dispatching mechanism again.

See [NDArrayOperatorsMixin](https://docs.scipy.org/doc/numpy-1.15.1/reference/generated/numpy.lib.mixins.NDArrayOperatorsMixin.html) for an example implementation.

shoyer commented on pandas/core/arrays/integer.py:
```diff
@@ -293,6 +295,44 @@ def __array__(self, dtype=None):
+        if (method == '__call__'
+                and ufunc.signature is None
+                and ufunc.nout == 1):
+            # only supports IntegerArray for now
```
This should check for `isinstance(.., IntegerArray)`?

shoyer commented on pandas/core/arrays/integer.py:
```diff
@@ -293,6 +295,44 @@ def __array__(self, dtype=None):
+        # for binary ops, use our custom dunder methods
+        result = ops.maybe_dispatch_ufunc_to_dunder_op(
+            self, ufunc, method, *inputs, **kwargs)
+        if result is not None:
```
Is there a reason why you use a different sentinel value `None` rather than Python's standard `NotImplemented`?

jorisvandenbossche replied:
Not a specific reason, just that I didn't return anything in the `maybe_dispatch_ufunc_to_dunder_op`, which means the result is None (and our own dunder ops will never return None). 
But can make it more explicit.

jorisvandenbossche committed:
first dispatch before getting underlying values (eg for Series[Period] underlying values are object array -> not always doing the correct thing)

shoyer commented on pandas/core/series.py:
```diff
@@ -623,6 +623,29 @@ def view(self, dtype=None):
+                ufunc, method, *inputs, **kwargs)
+        if result is NotImplemented:
+            raise TypeError("The '{0}' operation is not supported for "
+                            "dtype {1}.".format(ufunc.__name__, self.dtype))
```
Instead of all this (checking for `__array_ufunc__` and then the branch/error message), you could just reinvoke the public ufunc again (as I suggested for IntegerArray). That would also make me more confidence that the proper delegation is happening.

jorisvandenbossche replied:
Thanks, that sounds like a better idea :)

shoyer commented on pandas/core/series.py:
```diff
@@ -623,6 +623,29 @@ def view(self, dtype=None):
+        if result is NotImplemented:
+            raise TypeError("The '{0}' operation is not supported for "
+                            "dtype {1}.".format(ufunc.__name__, self.dtype))
+        return self._constructor(result, index=self.index,
```
Note that not every ufunc returns one result, so you should either handle or explicitly error for the other cases. See NDArrayOperatorsMixin for how to handle this.

jorisvandenbossche replied:
Out of curiosity, do you have the same problem with the current implementation of `__array_wrap__` ?

jorisvandenbossche replied:
To answer my own question: yes, at least for ufuncs that return a scalar:

```
In [33]: s = pd.Series([1, 2, 3])

In [34]: np.add.reduce(s)
Out[34]: 
0    6
1    6
2    6
dtype: int64

In [35]: pd.__version__
Out[35]: '0.23.4'
```

But for ufuncs that return multiple values it seems to work with pandas master (wrapping each return element I suppose).

shoyer commented on pandas/core/series.py:
```diff
@@ -623,6 +623,29 @@ def view(self, dtype=None):
         return self._constructor(self._values.view(dtype),
                                  index=self.index).__finalize__(self)
 
+    def __array_ufunc__(self, ufunc, method, *inputs, **kwargs):
```
Normally we recommend doing a type check and returning `NotImplemented` if you see an unknown type . If `__array_ufunc__` always tries to do the operation, you can get non-commutative operations, e.g., `a + b != b + a`.

For example, consider another library like xarray that implements its own version of a `pandas.Series`. `Series + DataArray` and `DataArray + Series` should consistently give either `Series`, `DataArray` or an error (probably the safest choice) -- they shouldn't just return a result of the same type as the first argument.

This might even be an issue for `Series + DataFrame` or `np.add(Series, DataFrame)` -- you probably want to delegate to the DataFrame in that case, not the Series.

jorisvandenbossche replied:
I was thinking somehow: the `__array_ufunc__` of the underlying values we call here is already doing the check, so I don't need to do it here again. 
But indeed, that will probably mess up such cases. 

The problem is that for now the `__array_ufunc__` uses our own implementations of all arithmetic/comparison operators, so I should somehow list here everything that all those different ops accept, which might be a bit annoying.

shoyer replied:
I doubt you need to list types dependent on the operators -- it's *probably* enough to list the types that can be used in valid operations with `pandas.Series`.

It's true that you would possibly need to recurse into extension arrays to figure this out all the valid types, and it may not be easy to enumerate all scalar types, e.g., if IPAddressExtensionArray can handle arithmetic with IPAddress, how could pandas.Series know about that?

jorisvandenbossche replied:
Yes, I think it are mainly the scalar values which might be hard to list (apart from those, we would probably accept Series, Index, array, and a bunch of array-likes (list, tuple, ..) for the Series case). 
But eg a Series with datetimelike values can already handle our own Timestamp, Timedelta and offset objects, numpy scalars, `datetime` standard library scalars, integer, ..), and as I understand those should all be listed here? 
(but anyhow, if we implement an `__array_ufunc__` for eg PeriodArray, we would need to do that there as well, so need to make this list in any case)

But what you mention about custom ExtensionArrays is then even another problem, that we don't know what they would accept. By "recurse into extension arrays", would you mean something like checking the `ExtensionArray._HANDLED_TYPES` (and then expecting EA authors to set this properly)? Or some other way to inspect?

shoyer replied:
> By "recurse into extension arrays", would you mean something like checking the `ExtensionArray._HANDLED_TYPES` (and then expecting EA authors to set this specific property)? Or some other way to inspect?

Yeah, that's about the best I can imagine. But I agree, it doesn't seem easy! There's no general way to detect what types another object's `__array_ufunc__` method can handle without actually calling it.

It is probably reasonable to trade-off correctness for generality here, but we should be cognizant that we are making this choice. Alternatively, we might maintain an explicit "blacklist" of types which pandas can't handle, though this gets tricky with non-required imports (e.g., `dask` and `xarray` objects).

TomAugspurger replied:
+1 to instructing (requiring?) ExtensionArray authors to implement `_HANDLED_TYPES` if they implement `__array_ufunc__`.

jorisvandenbossche replied:
@shoyer in light of the above, how would you define the handled types for an object dtyped Series, in which you in principle can put any python object you want? For numpy arrays, things like addition, work element-wise for object dtype, so the following works:

```
In [101]: class A():
     ...:     def __init__(self, x):
     ...:         self.x = x
     ...:     def __add__(self, other):
     ...:         return A(self.x + other.x)
     ...:     def __repr__(self):
     ...:         return "A(x={})".format(self.x)
     ...:          

In [102]: a = np.array([A(1), A(2)], dtype=object)

In [103]: np.add(a, A(3))
Out[103]: array([A(x=4), A(x=5)], dtype=object)

In [104]: np.add(pd.Series(a), A(3))
Out[104]: 
0    A(x=4)
1    A(x=5)
dtype: object
```

Doing a more strict check for allowed types, then the above will no longer work.

jorisvandenbossche replied:
How is numpy actually dealing with the above? It makes an exception for object dtype?

So two ideas that might be useful here:

* We can make an exception for object dtype, and in that case allow more unknown objects
* We could check for objects that implement `__array_ufunc__` themselves? If they have the attribute, and are not known to us (not in our `_HANDLED_TYPES` list, eg dask or xarray objects), we raise NotImplemented, otherwise we pass through. 
  That at least solves the problem for the more known objects that will probably have the interface, but of course not yet for custom container types that do not implement it.

shoyer replied:
These are good questions. In principle NumPy follows the rules described at http://www.numpy.org/neps/nep-0013-ufunc-overrides.html#behavior-in-combination-with-python-s-binary-operations but I think there is something missing for unknown scalars / object arrays...

shoyer replied:
See https://github.com/numpy/numpy/issues/12258 for a description of how NumPy does things.

TomAugspurger replied:
To make sure I understand this thread: @shoyer kicked it off by saying

> Normally we recommend doing a type check and returning NotImplemented if you see an unknown type

We can handle the scalar types of EAs by requiring that they implement a `_HANDLED_TYPEs`, which would include their scalar type. Then if all the (unboxed from Series / Index) inputs are in handled types, we proceed.

object-dtype complicates things. Did we decide what to do there? Skip the check?

jorisvandenbossche committed:
fix Categorical: disallow all ufunc apart from ops

jorisvandenbossche commented:
@shoyer Thanks for the feedback!

jorisvandenbossche committed:
simplify calling ufunc on underlying values

jorisvandenbossche committed:
fix categorical not existing ops

jorisvandenbossche committed:
np.positive not available for older numpy versions

jorisvandenbossche committed:
fix multiple return values

jorisvandenbossche committed:
skip IntegerArray tests for older numpy versions

jorisvandenbossche committed:
also deal with no return value

TomAugspurger commented:
Ah, I though the input was passed... Will take a look.

jorisvandenbossche committed:
clean-up debugging left-over

jreback commented on pandas/core/arrays/categorical.py:
```diff
@@ -1271,6 +1272,19 @@ def __array__(self, dtype=None):
 
+    def __array_ufunc__(self, ufunc, method, *inputs, **kwargs):
+        # for binary ops, use our custom dunder methods
+        result = ops.maybe_dispatch_ufunc_to_dunder_op(
```
shouldn't this be in the super class? e.g. this is a pretty general condition

jreback commented on pandas/core/ops.py:
```diff
@@ -2146,3 +2146,52 @@ def wrapper(self, other):
+
+
+def maybe_dispatch_ufunc_to_dunder_op(self, ufunc, method, *inputs, **kwargs):
+
```
can you add a doc-string

jreback commented on pandas/core/series.py:
```diff
@@ -623,6 +623,33 @@ def view(self, dtype=None):
+        if result is not NotImplemented:
+            return result
+
+        inputs = tuple(
```
a few comments here on what you are doing

shoyer commented:
> Question: do we want ufuncs to work on IntegerArray for older numpy versions? (that don't support yet `__array_ufunc__`?

You are of course welcome to try, but I don't think it's worth the effort. `__array_wrap__` is a pretty half-baked solution.

jreback commented on pandas/core/arrays/categorical.py:
```diff
@@ -1271,6 +1272,19 @@ def __array__(self, dtype=None):
+            self, ufunc, method, *inputs, **kwargs)
+        if result is not NotImplemented:
+            return result
+        # for all other cases, raise for now (similarly as what happens in
```
blank line here

jreback commented on pandas/core/arrays/integer.py:
```diff
@@ -293,6 +295,43 @@ def __array__(self, dtype=None):
+            masks = [a._mask for a in inputs]
+            result = ufunc(*args, **kwargs)
+            mask = np.logical_or.reduce(masks)
+            if result.dtype.kind in ('i', 'u'):
```
blank line

jreback commented on pandas/core/arrays/integer.py:
```diff
@@ -293,6 +295,43 @@ def __array__(self, dtype=None):
+                result[mask] = np.nan
+                return result
+
+        # fall back to array for other ufuncs
```
should this be a generic function that one can just call  (e.g. put it in ops.py and pass in self, ufuc, method, inputs, kwargs)? IOW is this called anywhere else

jreback commented on pandas/core/ops.py:
```diff
@@ -2146,3 +2146,52 @@ def wrapper(self, other):
+        'eq': '__eq__',
+        'ne': '__ne__',
+    }
+
```
can you give some comments on what is going on at various poinits here

jreback commented on pandas/core/series.py:
```diff
@@ -623,6 +623,33 @@ def view(self, dtype=None):
+        inputs = tuple(
+            x._values if isinstance(x, type(self)) else x
+            for x in inputs
+        )
```
e.g. I think you are calling the same here (as i commented above)

jreback commented:
@jorisvandenbossche worth updating for the rc?

jreback commented:
@jorisvandenbossche can you merge master and update

TomAugspurger committed:
TST: Additional tests for Series ufuncs

This adds a set of tests for ufuncs on Series. The goal is to
establish the correct behavior prior to implementing `Series.__array_ufunc__`.

There are two kinds of xfails right now

1. Series[Sparse] fails because `Series.__array_ufunc__` doesn't
   yet dispatch to `Series.array.__array_ufunc__`
2. `ufunc(series, series)` when the two series are unaligned. It's
   been determined that these should align, but isn't currently
   implemented.

TomAugspurger committed:
fixup release note

TomAugspurger committed:
fixup release note

TomAugspurger committed:
fixups

TomAugspurger committed:
remove stale comment

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into jorisvandenbossche-array-ufunc

TomAugspurger committed:
xfail ufunc(series, index)

TomAugspurger committed:
32-bit compat

(cherry picked from commit ba5cb55e511ea484bf4a63d4c489767e5bed87a2)

TomAugspurger committed:
wip

TomAugspurger committed:
fixup

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into series-array-ufunc

TomAugspurger committed:
more

TomAugspurger committed:
lint

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into jorisvandenbossche-array-ufunc

TomAugspurger committed:
Merge branch 'series-array-ufunc' into jorisvandenbossche-array-ufunc

TomAugspurger committed:
fixup! more

TomAugspurger committed:
remove dead code

TomAugspurger committed:
todos

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into jorisvandenbossche-array-ufunc

TomAugspurger commented:
Pushed a largeish change to `Series.__array_ufunc__`. The basic flow is

1. maybe dispatch to dunder op
2. maybe align input series (handling dataframes is a TODO; I think we want to define a `DataFrame.__array_ufunc__` and defer to that)
3. extract the arrays from the input Series / Index
4. dtype check for whether we can handle the op. This assumes that the underlying array defining `__array_ufunc__` defines a `_HANDLED_TYPES` tuple. We'll require that for ExtensionArray subclasses, which is the world I care about. If we see an unhandled type, we return NotImplemented. *Unless* we're a `Series[object]` dtype. This exception should be revisited.
5. Apply the ufunc
6. Wrap the result(s).

probably some cleanup to do, certainly some docs to write, but this makes sense to me.

TomAugspurger committed:
remove compat

TomAugspurger committed:
object dtype tests

TomAugspurger committed:
wip

TomAugspurger committed:
doc, types

TomAugspurger committed:
compat

TomAugspurger committed:
fixups

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into jorisvandenbossche-array-ufunc

TomAugspurger committed:
added matmul

TomAugspurger committed:
start docs

TomAugspurger committed:
compat

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into jorisvandenbossche-array-ufunc

TomAugspurger commented:
@shoyer I'm a bit confused about the errors on NumPy dev (actually, I think this happens on NumPy 1.16 as well; we just elevate warnings to errors for the NumPy dev build).

It comes from trying to compare a 0-d scalar `numpy.datetime64` to a `Series[datetime64[ns, tz]]`. We'd like pandas to raise a TypeError there.

The `Series == scalar` direction is fine

```pytb
In [12]:     dts = pd.Series(pd.date_range('2016', periods=2, tz='UTC'))
    ...:     other = np.datetime64('2016-01-01')
    ...:

In [13]: dts == other
Traceback (most recent call last):
  File "<ipython-input-13-1aa12ffb5b5f>", line 1, in <module>
    dts == other
  File "/Users/taugspurger/sandbox/pandas/pandas/core/ops.py", line 1855, in wrapper
    pd.DatetimeIndex)
  File "/Users/taugspurger/sandbox/pandas/pandas/core/ops.py", line 1330, in dispatch_to_index_op
    result = op(left_idx, right)
  File "/Users/taugspurger/sandbox/pandas/pandas/core/indexes/datetimelike.py", line 123, in wrapper
    result = op(self._data, maybe_unwrap_index(other))
  File "/Users/taugspurger/sandbox/pandas/pandas/core/arrays/datetimes.py", line 142, in wrapper
    self._assert_tzawareness_compat(other)
  File "/Users/taugspurger/sandbox/pandas/pandas/core/arrays/datetimes.py", line 700, in _assert_tzawareness_compat
    raise TypeError('Cannot compare tz-naive and tz-aware '
TypeError: Cannot compare tz-naive and tz-aware datetime-like objects
```

The reverse is raising a couple issues because NumPy apparently tries to do the operation again, after casting the Series to an ndarray?

```pytb
In [14]: other == dts
/Users/taugspurger/Envs/pandas-dev/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3296: FutureWarning: Converting timezone-aware DatetimeArray to timezone-naive ndarray with 'datetime64[ns]' dtype. In the future, this will return an ndarray with 'object' dtype where each element is a 'pandas.Timestamp' with the correct 'tz'.
        To accept the future behavior, pass 'dtype=object'.
        To keep the old behavior, pass 'dtype="datetime64[ns]"'.
  exec(code_obj, self.user_global_ns, self.user_ns)
/Users/taugspurger/.virtualenvs/pandas-dev/bin/ipython:1: DeprecationWarning: elementwise comparison failed; this will raise an error in the future.
  #!/Users/taugspurger/Envs/pandas-dev/bin/python
Traceback (most recent call last):
  File "<ipython-input-14-85ca43130dc6>", line 1, in <module>
    other == dts
  File "/Users/taugspurger/sandbox/pandas/pandas/core/ops.py", line 1855, in wrapper
    pd.DatetimeIndex)
  File "/Users/taugspurger/sandbox/pandas/pandas/core/ops.py", line 1330, in dispatch_to_index_op
    result = op(left_idx, right)
  File "/Users/taugspurger/sandbox/pandas/pandas/core/indexes/datetimelike.py", line 123, in wrapper
    result = op(self._data, maybe_unwrap_index(other))
  File "/Users/taugspurger/sandbox/pandas/pandas/core/arrays/datetimes.py", line 142, in wrapper
    self._assert_tzawareness_compat(other)
  File "/Users/taugspurger/sandbox/pandas/pandas/core/arrays/datetimes.py", line 700, in _assert_tzawareness_compat
    raise TypeError('Cannot compare tz-naive and tz-aware '
TypeError: Cannot compare tz-naive and tz-aware datetime-like objects
```

The first warning comes from pandas, doing `Series.__array__(dtype=None)` on a `datetime64[ns, tz]` Series.

The second is from NumPy.

shoyer commented:
@TomAugspurger I suspect the problem has to do with how you've implemented `__array_ufunc__` for `np.equal`, which I'm guessing ends up recursively calling `ndarray.__eq__` via `maybe_dispatch_ufunc_to_dunder_op`.

When interacting with others that define `__array_ufunc__`, NumPy defines `ndarray.__eq__` (which implements `==`) by calling corresponding NumPy ufuncs, something like:
```
class ndarray:
    def __eq__(self, other):
        # if other has a __array_ufunc__ method 
        return np.equal(self, other)
```
(`NDArrayOperatorsMixin` defines equivalent methods with the full details in pure Python, if you're curious)

Apologies if this was already clear to you. I'm still not sure exactly why you get this warning, but hopefully this helps put you on the right track to figuring it out...

TomAugspurger commented:
Thanks @shoyer. Here's a simpler example, using NDArrayOperatorMixin

The only special part is a bit of debug printing and

```python
        if any(x.ndim == 0 for x in inputs):
            raise TypeError('bad!')
```

```python
import numbers
import numpy as np


i = 0

class ArrayLike(np.lib.mixins.NDArrayOperatorsMixin):
    def __init__(self, value):
        self.value = np.asarray(value)

    # One might also consider adding the built-in list type to this
    # list, to support operations like np.add(array_like, list)
    _HANDLED_TYPES = (np.ndarray, numbers.Number)

    def __array_ufunc__(self, ufunc, method, *inputs, **kwargs):
        global i
        i += 1

        print(i, inputs)
        out = kwargs.get('out', ())
        for x in inputs + out:
            # Only support operations with instances of _HANDLED_TYPES.
            # Use ArrayLike instead of type(self) for isinstance to
            # allow subclasses that don't override __array_ufunc__ to
            # handle ArrayLike objects.
            if not isinstance(x, self._HANDLED_TYPES + (ArrayLike,)):
                return NotImplemented

        # Defer to the implementation of the ufunc on unwrapped values.
        inputs = tuple(x.value if isinstance(x, ArrayLike) else x
                       for x in inputs)

        if any(x.ndim == 0 for x in inputs):
            raise TypeError('bad!')

        if out:
            kwargs['out'] = tuple(
                x.value if isinstance(x, ArrayLike) else x
                for x in out)
        result = getattr(ufunc, method)(*inputs, **kwargs)

        if type(result) is tuple:
            # multiple return values
            return tuple(type(self)(x) for x in result)
        elif method == 'at':
            # no return value
            return None
        else:
            # one return value
            return type(self)(result)

    def __repr__(self):
        return '%s(%r)' % (type(self).__name__, self.value)


if __name__ == '__main__':
    a = ArrayLike([1, 2, 3])
    b = np.array(1)

    print(b == a)
```

Running that, I see

```pytb
1 (array(1), ArrayLike(array([1, 2, 3])))
/Users/taugspurger/sandbox/pandas/bar.py:60: DeprecationWarning: elementwise comparison failed; this will raise an error in the future.
  print(b == a)
2 (ArrayLike(array([1, 2, 3])), array(1))
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
~/sandbox/pandas/bar.py in <module>
     58     b = np.array(1)
     59
---> 60     print(b == a)

~/Envs/pandas-dev/lib/python3.7/site-packages/numpy/lib/mixins.py in func(self, other)
     23         if _disables_array_ufunc(other):
     24             return NotImplemented
---> 25         return ufunc(self, other)
     26     func.__name__ = '__{}__'.format(name)
     27     return func

~/sandbox/pandas/bar.py in __array_ufunc__(self, ufunc, method, *inputs, **kwargs)
     32
     33         if any(x.ndim == 0 for x in inputs):
---> 34             raise TypeError('bad!')
     35
     36         if out:

TypeError: bad!
```

So it seems that someone (NumPy? Python?) falls back to calling `eq(a, b)` when `eq(b, a)` fails? This doesn't happen with other ops like `>`.

jorisvandenbossche commented on doc/source/user_guide/computation.rst:
```diff
@@ -5,6 +5,53 @@
+
+:class:`Series` implements ``__array_ufunc__``, which allows it to work with NumPy's
+`universal functions <https://docs.scipy.org/doc/numpy/reference/ufuncs.html>`_
+to be applied.
```
something wrong with this sentence I think

jorisvandenbossche commented on doc/source/user_guide/computation.rst:
```diff
@@ -5,6 +5,53 @@
+   ser = pd.Series([1, 2, 3, 4])
+   np.exp(ser)
+
+Like other parts of of the library, pandas will automatically align labeled inputs
```
```suggestion
Like other parts of the library, pandas will automatically align labeled inputs
```

jorisvandenbossche commented on pandas/core/arrays/integer.py:
```diff
@@ -345,6 +346,51 @@ def __array__(self, dtype=None):
+        def reconstruct(x):
+            if np.isscalar(x):
+                # reductions.
+                if mask.any():
```
The `mask.any()` here means that we follow numpy behaviour and not skip NAs ?

TomAugspurger replied:
I guess so. This affects, e.g. `np.add.reduce`

The current implementing calls the ufunc on `self._data`, including the dummy `1` values for missing values, which makes the value here incorrect when there are missing values.  If we wanted to return a non-NA value here, we would need to handle `.reduce` entirely separate from the rest.

jorisvandenbossche replied:
In latest numpy there is a ``where`` keyword for this. But yeah, can't really use that yet (unless we decide to raise an error for integerarray for older numpy versions)

jorisvandenbossche replied:
This 'scalar' branch, is that this relevant if we raise above for 'reduce' methods? (I don't think any of the non-reduce ufuncs can return a scalar

jorisvandenbossche commented on pandas/core/series.py:
```diff
@@ -700,6 +701,72 @@ def view(self, dtype=None):
+            for s in alignable[1:]:
+                index |= s.index
+            inputs = [x.reindex(index) for x, t in zip(inputs, types)
+                      if issubclass(t, Series)]
```
should there be a `else x` in the end for if you have a mixture of alignable and non-alignable objects?

TomAugspurger replied:
Are there any ufuncs that have more than 2 inputs? This is under an `if len(alignables) > 1`, so it should only be hit when we have 2 series.

But yes, I'll add an `else x` there, even if it may not (currently) be hit.

jorisvandenbossche replied:
Ah, that's a good point :) 
Indeed they all only have one or two inputs (https://docs.scipy.org/doc/numpy/reference/ufuncs.html). 

A comment about that might be just as good then

jorisvandenbossche commented on pandas/core/series.py:
```diff
@@ -700,6 +701,72 @@ def view(self, dtype=None):
+            name = None
+
+        def construct_return(result):
+            return self._constructor(result,
```
Do we need to handle scalars here as well?

TomAugspurger replied:
Yeah... I'm a bit worried about all the issues we've had with `lib.is_scalar` :/

jorisvandenbossche reviewed (commented):
Thanks for picking this up! some quick comments

shoyer commented:
@TomAugspurger Thanks for the clear test case! I'm not sure if this is fixable, but I filed a NumPy bug nontheless: https://github.com/numpy/numpy/issues/13841

TomAugspurger committed:
ignore for numpydev

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into jorisvandenbossche-array-ufunc

jorisvandenbossche commented:
Regarding the NA behaviour, there is an active discussion on masked arrays on the numpy dev mailing list. I didn't read the whole thing, but it might be relevant to see what they think there how a masked array should behave for ufuncs, because in the end the IntegerArray is similar to a MaskedArray (if it can have deviating behaviour regarding skipping NAs)

TomAugspurger commented:
Yep, I'm following that discussion as well.

TomAugspurger committed:
handle reduce

jreback commented on doc/source/user_guide/computation.rst:
```diff
@@ -5,6 +5,53 @@
 ===================
 
+.. _computation.ufuncs:
+
```
should reference ``DataFrame interoperability with NumPy functions`` _dsintro.numpy_interop here, and/or remove the previous and reference this section.

jreback commented on doc/source/user_guide/computation.rst:
```diff
@@ -5,6 +5,53 @@
 
+.. _computation.ufuncs:
+
+Universal Functions
```
Numpy Universal Functions

jreback commented on doc/source/whatsnew/v0.25.0.rst:
```diff
@@ -787,6 +787,7 @@ Sparse
 - Introduce a better error message in :meth:`Series.sparse.from_coo` so it returns a ``TypeError`` for inputs that are not coo matrices (:issue:`26554`)
 - Bug in :func:`numpy.modf` on a :class:`SparseArray`. Now a tuple of :class:`SparseArray` is returned (:issue:`26946`).
 
```
adding a note here?

jreback commented on pandas/core/arrays/integer.py:
```diff
@@ -345,6 +346,51 @@ def __array__(self, dtype=None):
+                if mask.any():
+                    return np.nan
+                return x
+            if is_integer_dtype(x.dtype):
```
how does this handle an upcasted result, e.g. we do ``np.exp(pd.Series([1,2,3], dtype='Int64'))``? and this is now a float array, does it hit this machinery?

TomAugspurger replied:
Yeah, it returns a float64 series (this is tested).

jreback commented on pandas/core/ops.py:
```diff
@@ -2349,3 +2357,76 @@ def wrapper(self, other):
+    self: ArrayLike,
+    ufunc: Callable,
+    method: str,
+    *inputs,
```
are these typable? (inputs & kwargs)

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into jorisvandenbossche-array-ufunc

jreback commented on pandas/core/series.py:
```diff
@@ -700,6 +701,72 @@ def view(self, dtype=None):
     # ----------------------------------------------------------------------
     # NDArray Compat
+    _HANDLED_TYPES = (Index, ExtensionArray, np.ndarray, numbers.Number,
+                      list)  # what other builtins? array? deque? ...
```
Iterable/Sequence?

TomAugspurger replied:
I'm not sure whether including ABCs here is appropriate or not. That kind of puts us in the highest position, even if we may want to raise (e.g. `np.add(Series, Xarray.DataSet)`). I don't know what's best hear.

shoyer replied:
Instead of using `numbers.Number`, you could explicitly check for a lack of `__array_priority__` and `__array_ufunc__` attributes to indicate that something is a scalar / can be safely coerced.  That is basically what NumPy does.

jbrockmendel replied:
@jorisvandenbossche and I were discussing using `__array_priority__` the other day.  This seems less likely to accidentally leave things out than _HANDLED_TYPES.

jreback commented on pandas/core/series.py:
```diff
@@ -700,6 +701,72 @@ def view(self, dtype=None):
+    _HANDLED_TYPES = (Index, ExtensionArray, np.ndarray, numbers.Number,
+                      list)  # what other builtins? array? deque? ...
+
+    def __array_ufunc__(self, ufunc, method, *inputs, **kwargs):
```
can you type these args?

jreback commented on pandas/core/series.py:
```diff
@@ -700,6 +701,72 @@ def view(self, dtype=None):
+
+        # dtype check: can we do this, given the inputs?
+        # It's expected that the following classes defer to us when
+        # any Series is present in inputes.
```
inputs (typo)

TomAugspurger committed:
update

jreback reviewed (changes requested):
should there not be a default ``__array_ufunc__`` in ExtensionOpsMixin?

TomAugspurger committed:
fixups

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into jorisvandenbossche-array-ufunc

TomAugspurger committed:
raise for reduce

TomAugspurger committed:
more tests

TomAugspurger committed:
more tests

TomAugspurger commented:
I implemented `DecimalArray.__array_ufunc__` as a basic test for an EA defining it. Things seem to be working well.

jorisvandenbossche reviewed (commented):
additions look good!

TomAugspurger committed:
35 compat

TomAugspurger committed:
remove old test

jorisvandenbossche commented:
> Instead of using numbers.Number, you could explicitly check for a lack of __array_priority__ and __array_ufunc__ attributes to indicate that something is a scalar / can be safely coerced. That is basically what NumPy does.

@shoyer that would also solve what we discussed before about for object dtype not knowing what the handled types are? (currently we are skipping the type check for object dtype data)

jorisvandenbossche commented:
I think this is good enough to put in the RC

shoyer commented:
> > Instead of using numbers.Number, you could explicitly check for a lack of **array_priority** and **array_ufunc** attributes to indicate that something is a scalar / can be safely coerced. That is basically what NumPy does.
> 
> @shoyer that would also solve what we discussed before about for object dtype not knowing what the handled types are? (currently we are skipping the type check for object dtype data)

Yes, I think this would be an elegant way to handle this. Here's what I wrote up last October:
https://github.com/numpy/numpy/issues/12258#issuecomment-432858949

It's up to you whether you want to check for `__array_priority__` *and* `__array_ufunc__` attributes. The former is just a legacy thing now from NumPy's perspective (it's impossible to map operator priorities to integers outside of a single project), but it's probably still worth checking because classes that set `__array_priority__` were explicitly opting out of NumPy's automatic arithmetic operations that treat everything as a scalar. This is a good indication that they shouldn't be treated as scalars by pandas, either.

jorisvandenbossche commented:
Ah, thanks for the link to that issue. Forgot about that discussion.

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into jorisvandenbossche-array-ufunc

jreback commented on doc/source/development/extending.rst:
```diff
@@ -208,6 +208,28 @@ will
+
+As part of your implementation, we require that you
+
+1. Define a ``_HANDLED_TYPES`` attribute, a tuple, containing the types your
```
are we now recommeding setting ``__array_priority__`` ?

jreback commented on pandas/core/series.py:
```diff
@@ -704,6 +706,82 @@ def view(self, dtype=None):
+            **kwargs: Any
+    ):
+        # TODO: handle DataFrame
+        from pandas.core.internals.construction import extract_array
```
can you move this import to the top?

TomAugspurger replied:
No, circular import.

jreback commented on pandas/core/series.py:
```diff
@@ -704,6 +706,82 @@ def view(self, dtype=None):
+        # TODO: dataframe
+        alignable = [x for x, t in zip(inputs, types) if issubclass(t, Series)]
+
+        if len(alignable) > 1:
```
are we not *always* alignable? when would not be

TomAugspurger replied:
e.g. `np.add(Series, ndarray)` or `np.add(Series, Index)` will not align.

jreback commented on pandas/core/series.py:
```diff
@@ -704,6 +706,82 @@ def view(self, dtype=None):
+        handled_types = sum([getattr(x, '_HANDLED_TYPES', ()) for x in inputs],
+                            self._HANDLED_TYPES + (Series,))
+        any_object = any(getattr(x, 'dtype', None) == 'object' for x in inputs)
+        # defer when an unknown object and not object dtype.
```
blank line before comments

jreback reviewed (commented):
some small comments; ok to merge for the rc

TomAugspurger commented:
I'll update to incorporate https://github.com/pandas-dev/pandas/pull/23293#issuecomment-507056773 later today. Fixing the CI for pandas and dask first.

jreback commented:
also, is there any way we can pin numpy 1.17rc1 in one of the builds just to be sure (as saw some other breakages on this); even though we test master.

TomAugspurger committed:
fixup

TomAugspurger commented:
I tried to implement the array_priority / array_ufunc check. A bit rushed today so I may have messed something up.

The main wrinkle is that `Series.__array_ufunc__` *doesn't* want to return `NotImpelmented` for ExtensionArrays defining `__array_ufunc__`. We want them to defer to us, so that we get consistent behavior for binops between Series and EA (we take care of unboxing and calling the ufunc on the array).

jorisvandenbossche commented:
If I am understanding correctly, we should only check for the presence of `__array_ufunc__` / `__array_priority__` on an object if it is not one of the `_HANDLED_TYPES`. And for series, the EA is in the handled types?

TomAugspurger commented:
Ah, OK I missed that in the discussion. Just to make sure, we only just `Series._HANDLED_TYPES`? We don't do the recursive `getattr(input, '_HANDLED_TYPES', ())`?

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into jorisvandenbossche-array-ufunc

TomAugspurger committed:
Fixups

TomAugspurger commented:
FYI, all green here.

jreback commented:
thanks @jorisvandenbossche and @TomAugspurger very nice. would be even nicer to shake this down a bit in master......

jreback commented:
any open issues for this?

jorisvandenbossche commented:
> any open issues for this?

https://github.com/pandas-dev/pandas/issues/22798 is related, but need to check if that can be fully closed.

> Ah, OK I missed that in the discussion. Just to make sure, we only just Series._HANDLED_TYPES? We don't do the recursive getattr(input, '_HANDLED_TYPES', ())?

It *might* be that we should still check the `_HANDLED_TYPES` of underlying EAs. 
Some EAs will want to handle scalars that have a `__array_priority__` set (like our own scalars), although this is maybe a sign that we should not define `__array_priority__` on them ...

TomAugspurger commented:
On the scalar __array_priority__, we should be OK as long as its lower than
Series.__array_priority__?

On Tue, Jul 2, 2019 at 9:06 AM Joris Van den Bossche <
notifications@github.com> wrote:

> any open issues for this?
>
> #22798 <https://github.com/pandas-dev/pandas/issues/22798> is related,
> but need to check if that can be fully closed.
>
> Ah, OK I missed that in the discussion. Just to make sure, we only just
> Series._HANDLED_TYPES? We don't do the recursive getattr(input,
> '_HANDLED_TYPES', ())?
>
> It *might* be that we should still check the _HANDLED_TYPES of underlying
> EAs.
> Some EAs will want to handle scalars that have a __array_priority__ set
> (like our own scalars), although this is maybe a sign that we should not
> define __array_priority__ on them ...
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pandas-dev/pandas/pull/23293?email_source=notifications&email_token=AAKAOIW6L6Q3DGFMJG2A4ODP5NOFVA5CNFSM4F6UI2YKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODZBMLIA#issuecomment-507692448>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAKAOIRTUSX3D7LJONJBFQLP5NOFVANCNFSM4F6UI2YA>
> .
>

jorisvandenbossche commented:
It seems we use 1000 for Series and EAs, and 100 for Timestamp and Timedelta, so that should be OK then?

TomAugspurger commented:
That's my understanding.

On Tue, Jul 2, 2019 at 9:11 AM Joris Van den Bossche <
notifications@github.com> wrote:

> It seems we use 1000 for Series and EAs, and 100 for Timestamp and
> Timedelta, so that should be OK then?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pandas-dev/pandas/pull/23293?email_source=notifications&email_token=AAKAOISPBW64J332Y4FZDFTP5NOZNA5CNFSM4F6UI2YKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODZBM43Q#issuecomment-507694702>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAKAOIWW7S6SWRDDGKDR4PDP5NOZNANCNFSM4F6UI2YA>
> .
>
